File: credits/scripts/check-names.pl
#!/usr/bin/perl -w
#

use LWP::UserAgent;
use Text::CSV;
use IO::String;
use Data::Dumper;
use Text::Unaccent;

binmode STDOUT, ":utf8";
binmode STDERR, ":utf8";

my $csv = Text::CSV->new({binary => 1});
my @emails = ();

# Read inclusive range from command line
my ($from, $to) = @ARGV;
($from) || usage();
$to ||= $from;

$from =~ /^\d+$/ || usage();
$to =~ /^\d+$/ || usage();

if ($from > $to) {
    abort("Please specify a line number range, lowest number first.");
}

if ($from == 1) {
    abort("First line included in range, but first line is column headers.");
}

# Download file
my $ua = LWP::UserAgent->new;
$ua->cookie_jar({});

my $response = $ua->get("https://docs.google.com/spreadsheet/ccc?key=__KEY__&output=csv");

if (!$response->is_success) {
    abort("Couldn't get credits data from Google Docs.");
}

# Parse into CSV
my $credits_data_io = IO::String->new($response->decoded_content);

$csv->column_names("timestamp",
                   "name",
                   "sortkey",
                   "email",
                   "citation",
                   "evidence",
                   "notes");

$entries = $csv->getline_hr_all($credits_data_io);

# Make sure range is present
my $num_lines = scalar(@$entries);
if ($num_lines < $to) {
  abort("Range exceed size of sheet. Requested line $to but only $num_lines available.");
}

for (my $i = $from; $i <= $to; $i++) {
    my $entry = $entries->[$i - 1]; # Lines are 1-based, but array is 0-based

    my $name     = trim($entry->{'name'});
    my $sortkey  = trim($entry->{'sortkey'});
    my $email    = trim(lc($entry->{'email'}));
    my $citation = trim($entry->{'citation'});
    my $notes    = trim($entry->{'notes'});

    next if $notes ne "Y";

    # Deal with errant ' characters by stopping quote, starting new quote
    # with double quote characters, doing the single quote, then switching
    # back.
    # http://stackoverflow.com/questions/1250079/escaping-single-quotes-within-single-quoted-strings
    $citation =~ s/'/'"'"'/g;
    my $checkin  = $name . ' <' . $email . '>: "' . $citation . '"';

    if (!$sortkey) {
        # No specific sortkey given.
        # Names with > 2 parts default to assuming single family name for sort.
        $name =~ m/^(.*) ([^ ]+)$/;
        $sortkey = $2;
    }

    if (!$name || !$sortkey) {
        abort("Name or sortkey missing.");
    }

    # Make sortkey ASCII
    $sortkey = lc(unac_string('utf-8', $sortkey));

    # Check for duplicate
    open(my $NAMES, '<:encoding(utf8)', "../names.csv") || abort("Can't open names.csv");
    my %names;
    while (<$NAMES>) {
        chomp;
        my ($existing_name) = split(",");
        $names{$existing_name} = 1;
    }

    if ($names{$name}) {
        abort($name . " is a duplicate");
    }

    my $line = "$name,$sortkey";

    # Add name and commit
    system("echo $line >> ../names.csv");

    print "($sortkey) " . $checkin . "\n";

    system("git commit ../names.csv -m '$checkin'");

    push(@emails, $email);
}

print "Send email to: \n" . join(", ", @emails) . "\n\n";

sub usage {
    print "Usage: import <from> <to>\n";
    print "  where <from> and <to> are integer line numbers in the credits sheet.\n";
    exit(1);
}

sub trim {
    my ($string) = @_;
    $string =~ s/^\s+//;
    $string =~ s/\s+$//;
    return $string;
}

sub abort {
    print STDERR $_[0] . "\n";
    if (@emails) {
        print "Send email to: \n" . join(", ", @emails) . "\n\n";
    }

    exit(1);
}

File: credits/scripts/check-names.py
import csv
import requests
import sys
import git

def process_credits(lower_bound, upper_bound, names_to_be_added, names_added, emails, checkin, repo):
	response_data_url  = "https://docs.google.com/spreadsheet/ccc?key=__KEY__&output=csv"

	with requests.Session() as s:
		download = s.get(response_data_url)
		decoded_content = download.content.decode('utf-8')
		data = csv.reader(decoded_content.splitlines(), delimiter=',')
		for row in (r for i, r in enumerate(data) if lower_bound<=i<=upper_bound):
			if row[6] == 'Yes':
				data = "{name}:{sortkey}:{email}:{citation}".format(name=row[1], sortkey=row[2], email=row[3], citation=row[4])
				names_to_be_added.append(data)

	with open('../names.csv', 'r' , encoding='ISO-8859-1') as names:
		data_name = csv.reader(names)
		for to_add_name in names_to_be_added:
			flag = False
			for data in data_name:
				if data[0] == to_add_name.split(':')[0]:
					flag = True
					break
			if flag == False:
				emails.append(to_add_name.split(':')[2])
				names_added.append((',').join(to_add_name.split(':')[0:2]))
				checkin.append("{name} <{email}>: \"{citation}\"".format(name=to_add_name.split(':')[0], email=to_add_name.split(':')[2], citation=to_add_name.split(':')[3]))
			else:
				# If one entry is found to be existing in names.csv, it will simply ignore this and proceed with next one after printing this.
				print("Found duplicate name - {name}".format(name=to_add_name.split(':')[0]), " - Skipping this entry and proceeding.")

	for name, commit_message in zip(names_added, checkin):
		with open('../names.csv', 'a',encoding='utf-8') as names:
			names.write(name+'\n')
			names.close()
		repo.git.add('./credits/names.csv')
		repo.git.commit(m=commit_message)

	# Printing count and email ids at the end.
	print("\nTotal number of names added: ", len(emails))
	print("\nEmail IDs to send mails to: ", emails)
	print("\nNames of newly added contributors: ", names_added)

if __name__ == "__main__":
	repo = git.Repo('../../')
	lower_bound = sys.argv[1]
	upper_bound = sys.argv[2]
	names_to_be_added = []
	names_added = []
	emails = []
	checkin = []
	# lower_boud and upper_bound are substracted by 1 to ignore first heading row from excel file. 
	# Providing 2 & 3 as inputs to this script will process first & second entry of excel sheet as they are the ones in row 2 & 3 of excel sheet.
	process_credits(int(lower_bound)-1, int(upper_bound)-1, names_to_be_added, names_added, emails, checkin, repo)
